runtime/src/main/jni/v8_inspector/src/inspector/v8-page-agent-impl.cpp

```diff
@@ -0,0 +1,257 @@
+    V8InspectorSessionImpl* session, protocol::FrontendChannel* frontendChannel,
+    protocol::DictionaryValue* state)
+    : m_session(session), m_frontend(frontendChannel),
+      m_state(state),
+      m_enabled(false),
+      m_frameIdentifier("NSFrameIdentifier"),
+      m_frameUrl("file://") {}
+
+V8PageAgentImpl::~V8PageAgentImpl() {}
+
+void V8PageAgentImpl::enable(ErrorString*) {
+    if (m_enabled) {
+        return;
+    }
+
+    m_state->setBoolean(PageAgentState::pageEnabled, true);
+
+    m_enabled = true;
+}
+
+void V8PageAgentImpl::disable(ErrorString*) {
+    if (!m_enabled) {
+        return;
+    }
+
+    m_state->setBoolean(PageAgentState::pageEnabled, false);
+
+    m_enabled = false;
+}
+
+void V8PageAgentImpl::restore() {
+    if (!m_state->booleanProperty(PageAgentState::pageEnabled, false)) {
+        return;
+    }
+
+    ErrorString ignored;
+    enable(&ignored);
+}
+
+void V8PageAgentImpl::getResourceTree(
+    ErrorString*, std::unique_ptr<protocol::Page::FrameResourceTree>* out_frameTree) {
+    std::unique_ptr<protocol::Page::Frame> frameObject = protocol::Page::Frame::create()
+            .setId(m_frameIdentifier.c_str())
+            .setLoaderId("Loader Identifier")
+            .setMimeType("text/directory")
+            .setSecurityOrigin("")
+            .setUrl(m_frameUrl.c_str())
+            .build();
+
+    auto subresources = protocol::Array<protocol::Page::FrameResource>::create();
+
+    auto resources = v8_inspector::utils::PageResource::getPageResources();
+
+    for (auto const& resource : resources) {
+        auto pageResource = resource.second;
+        std::unique_ptr<protocol::Page::FrameResource> frameResource = protocol::Page::FrameResource::create()
+                .setUrl(pageResource.getFilePath())
+                .setType(pageResource.getType())
+                .setMimeType(pageResource.getMimeType())
+                .build();
+
+        subresources->addItem(std::move(frameResource));
+    }
+
+    *out_frameTree = protocol::Page::FrameResourceTree::create()
+                     .setFrame(std::move(frameObject))
+                     .setResources(std::move(subresources))
+                     .build();
+
+}
+
+void V8PageAgentImpl::getResourceContent(ErrorString* errorString, const String& in_frameId,
+        const String& in_url, String* out_content,
+        bool* out_base64Encoded) {
+    if (in_url.utf8().compare(m_frameUrl) == 0) {
+        *out_content = "";
+        *out_base64Encoded = false;
+
+        return;
+    }
+
+    std::map<std::string, v8_inspector::utils::PageResource> cachedPageResources = utils::PageResource::s_cachedPageResources;
+    if (utils::PageResource::s_cachedPageResources.size() == 0) {
+        cachedPageResources = utils::PageResource::getPageResources();
+    }
+
+    auto it = cachedPageResources.find(in_url.utf8());
+    if (it == cachedPageResources.end()) {
+        *errorString = "Resource not found.";
+
```

Comment:
Nice catch! The method would crash if the out_content parameter isn't set, so it needs to be set to an empty string/resource-not-found constant.